File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        set<vector<int>>hashSets;
        vector<vector<string>>ans;
        map<vector<int>,vector<string>>hashMap;
        for(auto it:strs){
            vector<int>hashSet(27,0);
            for(int i=0;i<it.size();i++)
                hashSet[int(it[i]-'a')]++;
            if(hashSets.find(hashSet)==hashSets.end()){
                hashSets.insert(hashSet);
            }
            hashMap[hashSet].push_back(it);
        }
        for(auto it:hashMap){
            ans.push_back(it.second);
        }
        return ans;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java


public class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> hashMap = new HashMap<>();
        List<List<String>> ans = new ArrayList<>();

        for (String str : strs) {
            char[] charArray = str.toCharArray();
            Arrays.sort(charArray);
            String sortedStr = new String(charArray);

            if (!hashMap.containsKey(sortedStr)) {
                hashMap.put(sortedStr, new ArrayList<>());
            }
            hashMap.get(sortedStr).add(str);
        }

        for (Map.Entry<String, List<String>> entry : hashMap.entrySet()) {
            ans.add(entry.getValue());
        }

        return ans;
    }
}
